0026-an-extensible-db-design.adoc

```diff
@@ -0,0 +1,180 @@
+
+In order to store details for each swap such as data we need for creating the HTLCs for each protocol we introduce `\{protocol\}_swap_details`, where `\{protocol\}` can be `han_bitcoin`, ``han_ethereum`, `herc20` or `halight` for now.
+Similar to the `Swaps` table, we add information which is only known at a later point of time into a sub-table.
+This resulted in quite an excess of sub-tables such as `han_bitcoin/ethereum_redeem_identity`, `han_bitcoin/ethereum_refund_identity`, etc.
+
+Some notable sub-tables:
+
+* `han_bitcoin/ethereum_swap_detail` contains data needed to create the HTLCs. We decided to give this table a generated primary key, rather than using `swap_local_id` as the primary key to allow same chain swaps.
+* `han_bitcoin_ethereum_redeem/refund_identity`: Since some identities are exchanged during announcement, the identities are kept in extension tables.
+
+One could optimize here and combine all ledger specific identities into an own table `han_bitcoin_identity` to hold all Bitcoin `refund` and `redeem` identities.
+This would mean though that we would need an additional unique id in those sub-tables, i.e. we can't use the `fk` as `pk`.
+
+* `han_bitcoin/ethereum_ledger_state`: is a new table which we introduced in this database schema.
+It records observed ledger events such as `fund`/`redeem`/`refund`-transaction.
+We extracted the states into `han_state` in order to reduce duplication in each table and not have an extra column for the event-type (`fund`/`redeem`/`refund`).
+This allows us to reuse these states among all `han_*` protocols.
+
+HErc20 and HaLight follow the same principles as described above.
+For completeness we depict `herc20` and `halight` diagrams below.
+
+image::http://www.plantuml.com/plantuml/proxy?src=https://raw.githubusercontent.com/comit-network/spikes/14e942f9c5ea1791f01bb65db0e09ac11cf5ab4c/assets/db_schema/herc20_swap_protocols.puml&fmt=svg[Herc20Protocols]
+
+image::http://www.plantuml.com/plantuml/proxy?src=https://raw.githubusercontent.com/comit-network/spikes/14e942f9c5ea1791f01bb65db0e09ac11cf5ab4c/assets/db_schema/halight_swap_protocols.puml&fmt=svg[HalightProtocols]
+
+=== Swap State
+
+image::http://www.plantuml.com/plantuml/proxy?src=https://raw.githubusercontent.com/comit-network/spikes/14e942f9c5ea1791f01bb65db0e09ac11cf5ab4c/assets/db_schema/swap_states.puml&fmt=svg[SwapStates]
+
+Additional to the specific `alpha` and `beta` ledger state (depicted in <<Protocols>>) a `swap` has two additional states:
+
+. the _communication_ state which depicts the state during the announcement phase, i.e. `sent`, `accepted`, `rejected`
+. the aggregated state of the swap, aggregation of the communication state, ledger states, plus an additional error state, i.e. `in_progress`, `swapped`, `not_swapped`, `internal_failure`
```

Comment:
I wouldn't track this in the DB, it is redundant data and can always be derived from everything else in the DB.

Reply:
We had this discussion as well and thought it would be more efficient to store in the DB than always derive it :)

Reply:
I think this is actually a bit more problematic. 
We cannot properly derive all `swap_states`, because of the `INTERNAL_FAILURE` state. This one is orthogonal to the other states. Hence, I think this should actually be tracked in the DB...?

Reply:
For efficiency reasons, it might be helpful to separate completed swaps (where the status will never change again) from ones that are in progress.

As long as data is changing, I would hesitate to store things that can run out of sync.

Reply:
I see your point, this adds another layer of complexity though.
If we consider separating swaps that are `IN_PROGRESS` from `SWAPPED`/`NOT_SWAPPED`/`INTERNAL_FAILURE` ones, would we do that with different tables?
If so, would we then remove a swap that is is a final state (`SWAPPED`/`NOT_SWAPPED`/`INTERNAL_FAILURE`) from the `IN_PROGRESS` table? (I suppose so.)

Reply:
I would just avoid the overall concept of an aggregated state in the database. I think it goes against basic SQL DB rules where you don't store derived/redundant data.

You can simply have a `completed_at` column on the swap and if you check that one, you know you don't need to spawn watchers but can assume all the state is locally available and not gonna change.

Reply:
I am not sure that properly solves the problem, as there are different final states and it might be relevant which final state the swap had. 
Otherwise the final state `INTERNAL_FAILURE` would never be recorded. 
I think we would have to optimise the state mode, otherwise we should store the "derived" state - to my mind `INTERNAL_FAILURE` is actually not derived, or am I wrong?

Reply:
My opinion: Remove `swap_state` from the DB for now, it was not in scope in the first place.
I'd keep `ledger_state` and `communication_state` as it is.

Reply:
> My opinion: Remove `swap_state` from the DB for now, it was not in scope in the first place.
> I'd keep `ledger_state` and `communication_state` as it is.

Fine with me, it is just more work than originally intended.

Reply:
I've removed `ledger_state` as this can be derived easily but kept `communication_state`.